### PR TITLE
Add sample order invariance to estimator_checks

### PR DESCRIPTION
**Pull request by anhqngo**
#### Reference Issues/PRs
Fixes #8695

#### What does this implement/fix? Explain your changes.
Added `check_methods_sample_order_variance` function, which checks for method invariance under sample order (i.e. results should not change when sample before or after applying the method).

#### Any other comments?
The original issue mentions using `assert_array_equal` to check for invariance. While this method works for `predict`, `decision_function`, `score_samples`, and `predict_proba`, it fails for `transform`. Therefore, I have opted for `assert_allclose_dense_sparse` with `atol=1e-9` instead of `assert_array_equal`, which passes the tests for all methods.

### CODE REVIEW

File: sklearn/utils/estimator_checks.py [hunk 1]
```diff
@@ -1124,6 +1125,41 @@ def check_methods_subset_invariance(name, estimator_orig):
                             atol=1e-7, err_msg=msg)
 
 
+@ignore_warnings(category=FutureWarning)
```
Comment (flosincapite) [inline sklearn/utils/estimator_checks.py:1128]: Looks good! As far as I can tell, this works as advertised.

File: sklearn/utils/estimator_checks.py [hunk 2]
```diff
@@ -1124,6 +1125,41 @@ def check_methods_subset_invariance(name, estimator_orig):
...
+def check_methods_sample_order_invariance(name, estimator_orig):
+    # check that method gives invariant results if applied
+    # on a subset with different sample order
+    rnd = np.random.RandomState(0)
```
Comment (flosincapite) [inline sklearn/utils/estimator_checks.py:1132]: I would recommend dependency-injecting this, but I see that this is the idiom within the rest of the code base.

File: sklearn/utils/estimator_checks.py [hunk 3]
```diff
@@ -1124,6 +1125,41 @@ def check_methods_subset_invariance(name, estimator_orig):
...
+    for method in ["predict", "transform", "decision_function",
+                   "score_samples", "predict_proba"]:
+        msg = ("{method} of {name} is not invariant when applied to a subset"
+               "with different sample order.").format(method=method, name=name)
+
+        if hasattr(estimator, method):
+            assert_allclose_dense_sparse(getattr(estimator, method)(X)[idx],
+                                         getattr(estimator, method)(X[idx]),
+                                         atol=1e-9,
+                                         err_msg=msg)
```
Comment (flosincapite) [inline sklearn/utils/estimator_checks.py:1160]: Notice that there are also [tests](https://github.com/MLH-Fellowship/scikit-learn/blob/efef74778644a6162c20dba9f26fcc5d24672bf8/sklearn/utils/tests/test_estimator_checks.py). In general, whenever you add a feature or alter code, a test somewhere should change as well. If you look in the linked file, you should see examples for how to add a test. Happy to discuss this with you if you're new to testing!

**Review by flosincapite — CHANGES_REQUESTED** [review]
Code looks good, but I want tests :).

**Comment by flosincapite** [comment]
Why don't you push your tests in their current state? I'd like to take a look!

**Comment by anhqngo** [comment]
@flosincapite alright I added my initial test. A couple of changes: I switched the order of the tests and now my test is being invoked, which is good. The code passes all tests locally now. 

Also, instead of random sampling from `X`, I just shuffled `X` around and then check for invariance since there's already `check_methods_subset_invariance` that takes care of the "subset" part.

File: sklearn/utils/tests/test_estimator_checks.py [hunk 4]
```diff
@@ -251,6 +251,30 @@ def predict(self, X):
...
+        return self
+
+    def predict(self, X):
+        X = check_array(X)
+        # check if the sample order is different from the original
+        # or not. If the sample order is different, then just return
+        # an array of zeros.
+        # TODO: For now, im just checking if the first instance in the sample
+        # is the same, but I will write a function to check this
+        if (X[0] != self.first).all():
```
Comment (flosincapite) [inline sklearn/utils/tests/test_estimator_checks.py]: Hmm, why all()? I would assume you'd want any--if any element of the sample has changed, then you know the sample order is different. I might be misunderstanding something, though.

File: sklearn/utils/tests/test_estimator_checks.py [hunk 5]
```diff
@@ -251,6 +251,30 @@ def predict(self, X):
...
+
+        self.first = X[0]
+        return self
+
+    def predict(self, X):
+        X = check_array(X)
+        # check if the sample order is different from the original
+        # or not. If the sample order is different, then just return
+        # an array of zeros.
+        # TODO: For now, im just checking if the first instance in the sample
```
Comment (flosincapite) [inline sklearn/utils/tests/test_estimator_checks.py]: Nice use of TODO! I don't know if you meant you're still working on this (which is fine!) OR that you want to submit in the current state. If the latter was your intention, I'd encourage you instead to address this TODO first.

**Review by flosincapite — CHANGES_REQUESTED** [review]

**Comment by flosincapite** [comment]
Looks cool! Feel free to send it upstream to scikit!

Also, if you were waiting on me, please let me know--I sometimes miss notifications if you resolve my comments. Don't wait on me for days--feel free to poke :)
